src/math_utils/convolution.py

```
@@ -0,0 +1,23 @@
+
+def apply_convolution(image, kernel):
+    # Validate input
+    if not isinstance(image, Image.Image):
+        raise TypeError("Input must be a PIL Image object")
+    if not isinstance(kernel, np.ndarray) or kernel.ndim != 2:
+        raise TypeError("Kernel must be a 2D NumPy array")
+        
+    image_array = np.array(image.convert("L"))  # grayscale
+    kernel_height, kernel_width = kernel.shape
+    # Calculate padding dynamically based on kernel size
+    pad_h = kernel_height // 2
```

Ruff (0.8.2)

Local variable `pad_h` is assigned to but never used

Remove assignment to unused variable `pad_h`

(F841)